src/planet-a/IcoSphereTree.cpp

```diff
@@ -501,13 +557,153 @@ TriangleSideTransform IcoSphereTree::transform_to_ancestor(
 
             out.m_scale *= 0.5f;
 
-            out.m_translation
-                    = out.m_translation * 0.5f
-                    + 0.5f * (side == (tri->m_siblingIndex + 1) % 3);
+            out.m_translation = out.m_translation * 0.5f
+                              + 0.5f * (side == (pTri->m_siblingIndex + 1) % 3);
+        }
+
+        curT = pTri->m_parent;
+    }
+
+}
+
+IcoSphereTreeObserver::Handle_T IcoSphereTree::event_add(
+        std::weak_ptr<IcoSphereTreeObserver> observer)
+{
+    return m_observers.emplace(m_observers.end(), std::move(observer));
+}
+
+void IcoSphereTree::event_remove(IcoSphereTreeObserver::Handle_T observer)
+{
+    // TODO
+}
+
+void IcoSphereTree::event_notify()
+{
+    if (!m_trianglesAdded.empty())
+    {
+        for (std::weak_ptr<IcoSphereTreeObserver> ob : m_observers)
+        {
+            // TODO: get rid of expired ones
+            ob.lock()->on_ico_triangles_added(m_trianglesAdded);
+        }
+        m_trianglesAdded.clear();
+    }
+
+    if (!m_trianglesRemoved.empty())
+    {
+        for (std::weak_ptr<IcoSphereTreeObserver> ob : m_observers)
+        {
+            // TODO: get rid of expired ones
+            ob.lock()->on_ico_triangles_removed(m_trianglesRemoved);
         }
+        m_trianglesRemoved.clear();
+    }
 
-        curT = tri->m_parent;
+    if (!m_vrtxRemoved.empty())
+    {
+        for (std::weak_ptr<IcoSphereTreeObserver> ob : m_observers)
```

Comment:
Iterating over the list 3 times is expensive.
Locking the shared_ptr 3 times is expensive.

Better to loop, and lock, once, and then do your if checks inside the loop and lock.